rust/noosphere-ns/src/bin/orb-ns/runner/runner.rs

```diff
@@ -0,0 +1,145 @@
+use crate::runner::config::{RunnerConfig, RunnerNodeConfig};
+use anyhow::Result;
+use futures::future::try_join_all;
+use noosphere_ns::{Multiaddr, NameSystem, NameSystemClient, PeerId};
+use noosphere_storage::{MemoryStorage, SphereDb};
+use serde::Serialize;
+use std::{
+    future::Future,
+    net::{SocketAddr, TcpListener},
+    pin::Pin,
+    sync::Arc,
+    task,
+};
+use tokio::sync::Mutex;
+use url::Url;
+
+#[cfg(feature = "api-server")]
+use noosphere_ns::server::APIServer;
+
+#[cfg(not(feature = "api-server"))]
+struct APIServer;
+#[cfg(not(feature = "api-server"))]
+impl APIServer {
+    pub fn serve(_ns: Arc<Mutex<NameSystem>>, _listener: TcpListener) -> Self {
+        APIServer {}
+    }
+}
+
+/// Wrapper around [NameSystem], and optionally an APIServer, and useful
+/// static info to log to the user.
+#[derive(Serialize)]
+struct ActiveNameSystem {
+    #[serde(skip_serializing)]
+    name_system: Arc<Mutex<NameSystem>>,
+    peer_id: PeerId,
+    listening_address: Option<Multiaddr>,
+    api_address: Option<Url>,
+    #[serde(skip_serializing)]
+    _api_thread: Option<APIServer>,
+}
+
+impl ActiveNameSystem {
+    pub(crate) async fn try_from_config(mut config: RunnerNodeConfig) -> Result<Self> {
+        let store = SphereDb::new(&MemoryStorage::default()).await?;
+        let node = NameSystem::new(&config.key_material, store, config.dht_config.to_owned())?;
+        let peer_id = node.peer_id().to_owned();
+
+        let listening_address = if let Some(requested_addr) = config.listening_address.take() {
+            // Request address from DHT to resolve default port (0) to
+            // selected port.
+            let resolved_addr = node.listen(requested_addr.to_owned()).await?;
+            Some(resolved_addr)
+        } else {
+            None
+        };
+
+        node.add_peers(config.peers.to_owned()).await?;
+
+        let wrapped_node = Arc::new(Mutex::new(node));
+
+        let (api_address, api_thread) = if cfg!(feature = "api-server") {
+            if let Some(requested_addr) = config.api_address.take() {
+                let api_listener = TcpListener::bind(requested_addr)?;
+                let api_address = socket_addr_to_url(api_listener.local_addr()?)?;
+                (
+                    Some(api_address),
+                    Some(APIServer::serve(wrapped_node.clone(), api_listener)),
+                )
+            } else {
+                (None, None)
+            }
+        } else {
+            (None, None)
```

Comment:
Suggestions on collapsing these conditionals?

Reply:
In the isolation of this condition: I think this might be the best you can do. But, it is possible that a different structure in the related constructs could make this cleaner to read :shrug: